File: artifacts/api-server/src/services/integration-manager.ts
import { logger } from "../lib/logger";
import { integrationConfig } from "../lib/integration-config";

export interface ConnectorHealth {
  name: string;
  enabled: boolean;
  connected: boolean;
  status: "CONFIGURED" | "NOT_CONFIGURED" | "NOT_CONNECTED" | "UNAVAILABLE";
  requiresAuthorization: boolean;
  message: string;
}

export interface SubmitCaseResponse {
  status: "success" | "error";
  externalId?: string;
  systemName: string;
  error?: string;
}

export interface GetCaseStatusResponse {
  status: "success" | "error";
  externalId: string;
  externalStatus?: string;
  systemName: string;
  error?: string;
}

class IntegrationManager {
  private connectors: Map<string, any> = new Map();

  constructor() {
    this.initializeConnectors();
  }

  private initializeConnectors(): void {
    if (integrationConfig.ncrp.enabled) {
      logger.info("NCRP connector enabled");
      this.connectors.set("ncrp", integrationConfig.ncrp);
    }
    if (integrationConfig.sahyog.enabled) {
      logger.info("SAHYOG connector enabled");
      this.connectors.set("sahyog", integrationConfig.sahyog);
    }
    if (integrationConfig.vasp.enabled) {
      logger.info("VASP connector enabled");
      this.connectors.set("vasp", integrationConfig.vasp);
    }
  }

  async healthCheck(): Promise<{ integrations: ConnectorHealth[]; timestamp: string }> {
    const results: ConnectorHealth[] = [];

    // NCRP status
    const ncrpEnabled = integrationConfig.ncrp.enabled;
    results.push({
      name: "ncrp",
      enabled: ncrpEnabled,
      connected: false,
      status: ncrpEnabled ? "NOT_CONNECTED" : "NOT_CONFIGURED",
      requiresAuthorization: true,
      message: ncrpEnabled
        ? "NCRP configuration present but authorized API access is not available"
        : "NCRP requires authorized API credentials. Not configured.",
    });

    // SAHYOG status
    const sahyogEnabled = integrationConfig.sahyog.enabled;
    results.push({
      name: "sahyog",
      enabled: sahyogEnabled,
      connected: false,
      status: sahyogEnabled ? "NOT_CONNECTED" : "NOT_CONFIGURED",
      requiresAuthorization: true,
      message: sahyogEnabled
        ? "SAHYOG configuration present but authorized API access is not available"
        : "SAHYOG requires authorized API credentials. Not configured.",
    });

    // VASP status
    const vaspEnabled = integrationConfig.vasp.enabled;
    const vaspConnected = vaspEnabled && !!integrationConfig.vasp.apiUrl;
    results.push({
      name: "vasp",
      enabled: vaspEnabled,
      connected: vaspConnected,
      status: vaspEnabled
        ? vaspConnected
          ? "CONFIGURED"
          : "NOT_CONNECTED"
        : "NOT_CONFIGURED",
      requiresAuthorization: false,
      message: vaspEnabled
        ? vaspConnected
          ? "VASP provider configured"
          : "VASP configuration incomplete"
        : "VASP provider not configured",
    });

    return { integrations: results, timestamp: new Date().toISOString() };
  }

  async submitCase(
    systemName: string | string[],
    caseData: Record<string, unknown>
  ): Promise<SubmitCaseResponse> {
    const system = Array.isArray(systemName) ? systemName[0] : systemName;

    // NCRP and SAHYOG are never connected - return explicit error
    if (system === "ncrp" || system === "sahyog") {
      return {
        status: "error",
        systemName: system,
        error: `${system.toUpperCase()} integration is not available. Authorized API access is required but not configured. No external ID has been generated.`,
      };
    }

    if (!this.connectors.has(system)) {
      return {
        status: "error",
        systemName: system,
        error: `Integration ${system} not available. No external ID has been generated.`,
      };
    }

    // No real external submission is possible without authorized API access
    return {
      status: "error",
      systemName: system,
      error: `${system.toUpperCase()} integration is not available. Authorized API access is required but not configured. No external ID has been generated.`,
    };
  }

  async getCaseStatus(
    systemName: string | string[],
    externalId: string | string[]
  ): Promise<GetCaseStatusResponse> {
    const system = Array.isArray(systemName) ? systemName[0] : systemName;
    const extId = Array.isArray(externalId) ? externalId[0] : externalId;

    // NCRP and SAHYOG are never connected
    if (system === "ncrp" || system === "sahyog") {
      return {
        status: "error",
        systemName: system,
        externalId: extId,
        error: `${system.toUpperCase()} integration is not available. Authorized API access is required but not configured.`,
      };
    }

    if (!this.connectors.has(system)) {
      return {
        status: "error",
        systemName: system,
        externalId: extId,
        error: `Integration ${system} not available`,
      };
    }

    // No real external status query is possible without authorized API access
    return {
      status: "error",
      systemName: system,
      externalId: extId,
      error: `${system.toUpperCase()} status query is not available. Authorized API access is required but not configured.`,
    };
  }

  getEnabledConnectors(): string[] {
    return Array.from(this.connectors.keys());
  }
}

export const integrationManager = new IntegrationManager();

File: lib/integration_manager.py
"""Integration manager for CASHNET.

Manages NCRP, SAHYOG, and VASP connectors.
NCRP and SAHYOG are NEVER connected without authorized credentials.
"""

from __future__ import annotations

import logging
from typing import Any

from lib.integration_config import IntegrationConfig

logger = logging.getLogger(__name__)


class IntegrationManager:
    def __init__(self):
        self.config = IntegrationConfig()
        self.connectors: dict[str, Any] = {}

        if self.config.ncrp.enabled:
            self.connectors["ncrp"] = self.config.ncrp
        if self.config.sahyog.enabled:
            self.connectors["sahyog"] = self.config.sahyog
        if self.config.vasp.enabled:
            self.connectors["vasp"] = self.config.vasp

    def health_check(self) -> dict[str, Any]:
        results = []

        # NCRP status
        results.append(
            {
                "name": "ncrp",
                "enabled": self.config.ncrp.enabled,
                "connected": False,
                "status": (
                    "NOT_CONNECTED" if self.config.ncrp.enabled else "NOT_CONFIGURED"
                ),
                "requires_authorization": True,
                "message": "NCRP requires authorized API access. Not connected.",
            }
        )

        # SAHYOG status
        results.append(
            {
                "name": "sahyog",
                "enabled": self.config.sahyog.enabled,
                "connected": False,
                "status": (
                    "NOT_CONNECTED" if self.config.sahyog.enabled else "NOT_CONFIGURED"
                ),
                "requires_authorization": True,
                "message": "SAHYOG requires authorized API access. Not connected.",
            }
        )

        # VASP status
        vasp_connected = self.config.vasp.enabled and bool(self.config.vasp.api_url)
        results.append(
            {
                "name": "vasp",
                "enabled": self.config.vasp.enabled,
                "connected": vasp_connected,
                "status": "CONFIGURED" if vasp_connected else "NOT_CONFIGURED",
                "requires_authorization": False,
                "message": "VASP provider" if vasp_connected else "VASP not configured",
            }
        )

        return {"integrations": results}

    def submit_case_sync(
        self, system_name: str, case_data: dict[str, Any]
    ) -> dict[str, Any]:
        """Synchronous case submission (for testing)."""
        if system_name in ("ncrp", "sahyog"):
            return {
                "status": "error",
                "systemName": system_name,
                "error": f"{system_name.upper()} integration is not available. Authorized API access is required but not configured.",
            }

        if system_name not in self.connectors:
            return {
                "status": "error",
                "systemName": system_name,
                "error": f"Integration {system_name} not available. No external ID has been generated.",
            }

        # No real external submission is possible without authorized API access
        return {
            "status": "error",
            "systemName": system_name,
            "error": f"{system_name.upper()} integration is not available. Authorized API access is required but not configured. No external ID has been generated.",
        }

    def get_enabled_connectors(self) -> list[str]:
        return list(self.connectors.keys())
